Replace OpenTK with SDL 2 for Joystick, Gamepad, GamePlatform, GameWindow, Mouse and Keyboard

cra0zy opened the pull request:
Notes:
 - this does brake Angle platform

tomspilman commented:
Looks like you missed a file:

```
Cannot copy MonoGame.Framework/Desktop/MonoGame.Framework.dll.config to MonoGame.Framework/bin/Linux/AnyCPU/Release/MonoGame.Framework.config, as the source file doesn't exist.
```

cra0zy commented:
Ok, this breaks Mac as well... will put Mac Joystick to Default so it can compile until OpenTK replacement is finished.

cra0zy commented:
Few more notes:
 - added Gamepad here since I got it done before this got merged
 - vibrations are finally working \o/ (OpenTK didn't have them implemented)
   - I was planing on adding just rumble effect but thanks so me forgetting the SDL_INIT_HAPTIC flag I ended up going over @flibitijibibo old monogame-sdl code trying to figure out what I did wrong... and there I noticed how to setup LeftRight effect... 
 - This also adds sdl2 gamepad database file from: https://github.com/gabomdq/SDL_GameControllerDB

cra0zy commented:
More notes:
 - Updates GameWindow and GamePlatform to SDL2 backend
 - Mouse and Keyboard input is no longer working, I still have to update those

cra0zy commented:
And EVEN MORE MORE notes:
 - in order to calculate window titlebar size and border I had to make mouse start in center of window, hopefully that is ok

tomspilman commented:
> in order to calculate window titlebar and border size I had to make mouse start in center of window

Why is that?  Are we already running into problems with the SDL backend?

tomspilman commented on MonoGame.Framework/Game.cs:
```diff
@@ -19,6 +19,10 @@ namespace Microsoft.Xna.Framework
     public class Game : IDisposable
     {
+#if DESKTOPGL
+        public static bool Resizable = false;
```
First let's not add more platform specific code into `Game`.

Second isn't this what `GameWindow.AllowUserResizing` is for?

cra0zy replied:
SDL does not have a way of setting window resizable parameter after the window has already been created.

tomspilman replied:
> SDL does not have a way of setting window resizable parameter after the window has already been created.

We will have to find a way around that.  At least on startup we could delay creation of the SDL window until `Game.Initialize()` which currently when the window is actually shown.

But changing that after startup... including making the window borderless...  is all stuff we need to support.

Regardless of all that we won't be adding a `Game.Resizable` property.

cra0zy replied:
> We will have to find a way around that. At least on startup we could delay creation of the SDL window until Game.Initialize() which currently when the window is actually shown.
> 
> But changing that after startup... including making the window borderless... is all stuff we need to support.
> 
> Regardless of all that we won't be adding a Game.Resizable property.

Window creation is tied to a valid GL window in: https://github.com/cra0zy/MonoGame/blob/sdl2/MonoGame.Framework/Graphics/GraphicsDevice.OpenGL.cs#L100 I though doing this was more sensible for the moment.

Window can already be set borderless without problem, it's only Resizable property that is not setable after creation.

tomspilman replied:
> was more sensible for the moment.

Maybe for the moment... but we definitely don't want to leave window creation in `GraphicsDevice`.

> it's only Resizable property that is not setable after creation.

We can maybe make due without that for now, but we should talk to the SDL team about it.  If OpenTK supported it I don't see why SDL couldn't.

cra0zy replied:
> We can maybe make due without that for now, but we should talk to the SDL team about it. If OpenTK supported it I don't see why SDL couldn't.

OpenTK does not support it with SDL backend, that is why we kept SDL disabled for Linux in OpenTK.dll.config

> Maybe for the moment... but we definitely don't want to leave window creation in GraphicsDevice.

Window is not getting created in GraphicDevice, it's just that we need Window handle to initialize some GL stuff...

Anyway I can move window creation code in there and it should theoretically allow what you wanted with resizable parameter, do you want me to do that?

tomspilman replied:
> OpenTK does not support it with SDL backend,

I'm taking about OpenTK ... not OpenTK with SDL.  Did OpenTK support changing the window to resizable after creation?

cra0zy replied:
> I'm taking about OpenTK ... not OpenTK with SDL. Did/does OpenTK support changing the window to resizable after creation?

You don't get how opentk functions, you can't ask "does OpenTK support this" since it depends on the backend it runs with:
1) OpenTK has 2 backends for each platform, native and SDL
2) During initialization it will pick one (this can be forced within code)
3) For resizable, If it picked SDL it won't support it, if it picked native it will support it

All an all OpenTK can't help SDL with this. 

We can't use OpenTK to make SDL window resizable.

PS. SDL 1.2 did support changing resizable parameter after window creation.

tomspilman replied:
> We can't use OpenTK to make SDL window resizable.

You have completely misunderstood me.

Obviously SDL can't be fixed by OpenTK...   I was just trying to figure out if the problem existed in OpenTK (with native backend).

> PS. SDL 1.2 did support changing resizable parameter after window creation.

So that settles it...  SDL2 lost this feature and it is totally possible to change the resize window state on Mac and Linux platforms after window creation.

This was all I was trying to figure out.

tomspilman commented on MonoGame.Framework/SDL/SDLGameWindow.cs:
```diff
@@ -0,0 +1,197 @@
+                    SDL.SDL_GetWindowPosition(Handle, out x, out y);
+
+                    x += BorderX;
+                    y += BorderY;
```
I suspect this border calculation is unnecessary.

The docs say that `SDL_GetWindowSize()` return the size the client area... not the window size.

Also this post confirms it...

http://stackoverflow.com/a/31996739

Are you certain this is needed?

cra0zy replied:
I need border in order to return correct position, not size.

cra0zy commented on MonoGame.Framework/Input/Mouse.cs:
```diff
@@ -139,20 +94,21 @@ public static MouseState GetState(GameWindow window)
+            var pos = PrimaryWindow.Position;
+
+            window.MouseState.X = x - pos.X - BorderX;
+            window.MouseState.Y = y - pos.Y - BorderY;
```
@tomspilman This is another position where I needed border size, the reason why it's needed here is so we can track global mouse.

We can track mouse while it's inside window with `SDL.SDL_GetMouseState`, but as soon as it gets out it won't be tracked.

cra0zy commented:
Ok, so I decided to use `GetMouseState` for now, until I can setup a proper border detection code.

We are gonna set it up per platform :(

cra0zy commented:
@tomspilman I think I did what you wanted with the last commit.

cra0zy commented:
This is ready for merging.

WardBenjamin commented:
No it's not! Fix in a few minutes.

tomspilman commented:
Note I plan to be a lot more flexible with merges to this `replace_opentk` branch as I expect it to be all very experimental until we decide to merge it back to develop.

WardBenjamin commented:
@tomspilman @cra0zy  check out https://github.com/cra0zy/MonoGame/pull/1. Simple typo :P

cra0zy commented:
> @tomspilman @cra0zy check out cra0zy#1. Simple typo :P

So you made a PR for another PR instead of just warning me about the typo? #logic

Anyway I rebased and fixed the original commit.

amerkoleci commented:
I think would be great to distribuite also native SDL2 dlls for Windows, Linux and OSX.
Note that SDL 2.0.4 has support for Windows10 Universal app too, so probably can be used there too.

cra0zy commented:
> I think would be great to distribuite also native SDL2 dlls for Windows, Linux and OSX.

I agree with you on that one, but that update will come in a later PR that modifies project templates. If you are interested about general discussion on this topic look at #4399

> Note that SDL 2.0.4 has support for Windows10 Universal app too, so probably can be used there too.

Didn't know that, tho I wouldn't use SDL for it since we do get a higher level of control with the existing method. Plus it's much more complex using SDL on non Desktop stuff.

tomspilman commented:
> I wouldn't use SDL for it since we do get a higher level of control with the existing method.

Yeah... Win10 works great with existing MS APIs.  We really only need SDL for the GL desktop platforms.

cra0zy commented:
@tomspilman Can this be merged? I want to start working on some other parts of this(probably audio).

tomspilman commented on MonoGame.Framework/SDL/SDL2.cs:
```diff
@@ -0,0 +1,405 @@
+        public static extern int SDL_ShowCursor(int toggle);
+
+        [DllImport(nativeLibName, CallingConvention = CallingConvention.Cdecl, EntryPoint = "SDL_WarpMouseInWindow")]
+        public static extern void SDL_WarpMouseInWindow(IntPtr window, int x, int y);
```
Looks like you missed removing the `SDL_` prefix on these imports?

cra0zy replied:
Yup, fixed.

tomspilman commented:
Seems good to merge after the build finishes.  Since it is only being merged to the `replace_opentk` branch it won't break anything for the 3.5 release and is pretty low risk right now.

tomspilman commented:
Merged @cra0zy !
